### PR TITLE
Add GPULimits for vertex inputs

### LINKED ISSUE #693: Figure out the base vertex input limits
We've been asked to clarify the baseline for:
  - max number of vertex buffers
  - max number of vertex attributes
  - max attribute offset
  - max stride

### PR DESCRIPTION
**Pull request by jespertheend**
Fixes #693.

***
<a href="https://pr-preview.s3.amazonaws.com/jespertheend/gpuweb/pull/1274.html" title="Last updated on Dec 5, 2020, 12:43 AM UTC (1fa2bba)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/1274/df95e0f...jespertheend:1fa2bba.html" title="Last updated on Dec 5, 2020, 12:43 AM UTC (1fa2bba)">Diff</a>

### CODE REVIEW

**Comment by jespertheend** [comment]
Hmm, do I need IE status for these checks to pass? I have filled out the Invited Expert application form, but I'm not sure if this is necessary.

File: spec/index.bs [hunk 1]
```diff
@@ -1045,6 +1045,21 @@ a better limit is not specified.
...
+
+    <tr><td><dfn>maxVertexBuffers</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>8
+    <tr class=row-continuation><td colspan=4>
+        The maximum number of {{GPUVertexStateDescriptor/vertexBuffers}} when creating a {{GPURenderPipeline}}.
+
+    <tr><td><dfn>maxVertexAttributes</dfn>
+        <td>{{GPUSize32}} <td>Higher <td>16
+    <tr class=row-continuation><td colspan=4>
+        The maximum number of {{GPUVertexBufferLayoutDescriptor/attributes}} when creating a {{GPURenderPipeline}}.
```
Comment (kvark) [inline spec/index.bs:1057]: this should say "in total across  {{GPUVertexStateDescriptor/vertexBuffers}}" somewhere

File: spec/index.bs [hunk 2]
```diff
@@ -4171,10 +4189,10 @@ dictionary GPUVertexAttributeDescriptor {
 
     Return `true`, if and only if, all of the following conditions are satisfied:
 
-        1. |descriptor|.{{GPUVertexBufferLayoutDescriptor/attributes}}.length is less than or equal to 16.
-        1. |descriptor|.{{GPUVertexBufferLayoutDescriptor/arrayStride}} is less then or equal to 2048.
+        1. |descriptor|.{{GPUVertexBufferLayoutDescriptor/attributes}}.length is less than or equal to {{GPULimits/maxVertexAttributes}}.
```
Comment (kvark) [inline spec/index.bs:4192]: this check should be replaced by a combined check over the whole pipeline (as opposed to a single vertex buffer)

File: design/Limits.md [hunk 3]
```diff
@@ -14,16 +14,22 @@ dictionary GPULimits {
     unsigned long maxStorageBuffersPerShaderStage = 4;
     unsigned long maxStorageTexturesPerShaderStage = 4;
     unsigned long maxUniformBuffersPerShaderStage = 12;
+    unsigned long maxVertexBuffers = 8;
+    unsigned long maxVertexAttributes = 16;
+    unsigned long maxVertexArrayStride = 2048;
```
Comment (kvark) [inline design/Limits.md:19]: I'm not sure if `maxVertexArrayStride` is the name we want, but the editors can follow-up on that separately, no need to block this PR.

**Review by kvark — COMMENTED** [review]
Thank you for the PR!
There is a few things I believe need fixing, overall looks good!

**Comment by kainino0x** [comment]
> Hmm, do I need IE status for these checks to pass? I have filled out the Invited Expert application form, but I'm not sure if this is necessary.

The ipr pass doesn't have to check for us to land this (it's non-required). That said, I think we could safely mark this contribution as non-substantive.

**Comment by kainino0x** [comment]
Adding label for the limit name

**Review by kainino0x — COMMENTED** [review]
No additional comments :)

File: spec/index.bs [hunk 4]
```diff
@@ -4193,16 +4213,18 @@ Issue(https://github.com/gpuweb/gpuweb/issues/693): add a limit to the number of
...
     Return `true`, if and only if, all of the following conditions are satisfied:
 
-        1. |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}}.length is less than or equal to 8
+        1. |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}}.length is less than or equal to
+            {{GPULimits/maxVertexBuffers|GPULimits.maxVertexBuffers}}
         1. Each |vertexBuffer| layout descriptor in the list |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}}
             passes [$validating GPUVertexBufferLayoutDescriptor$](|vertexBuffer|, |vertexStage|)
+        1. For every |vertexBuffer| in |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}},
+            the sum of every |vertexBuffer|.{{GPUVertexBufferLayoutDescriptor/attributes}}.length combined
+            is less than or equal to {{GPULimits/maxVertexAttributes|GPULimits.maxVertexAttributes}}.
```
Comment (jespertheend) [inline spec/index.bs:4222]: Not sure if there's a better way to write this down but this is the best I've got.

File: design/Limits.md (hunk 3 above)
Comment (jespertheend) [reply design/Limits.md:19]: Should I create a new issue for this?

**Review by kvark — APPROVED** [review]

File: design/Limits.md (hunk 3 above)
Comment (kainino0x) [reply design/Limits.md:19]: We can file an issue during the Monday editing meeting if we don't immediately decide on something.

File: spec/index.bs (hunk 4 above)
Comment (kainino0x) [reply spec/index.bs:4222]: I'll tweak the grammar but this looks good.

File: spec/index.bs [hunk 5]
```diff
@@ -4209,9 +4213,13 @@ dictionary GPUVertexAttributeDescriptor {
...
     Return `true`, if and only if, all of the following conditions are satisfied:
 
-        1. |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}}.length is less than or equal to {{GPULimits/maxVertexBuffers}}
+        1. |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}}.length is less than or equal to
+            {{GPULimits/maxVertexBuffers|GPULimits.maxVertexBuffers}}
         1. Each |vertexBuffer| layout descriptor in the list |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}}
             passes [$validating GPUVertexBufferLayoutDescriptor$](|vertexBuffer|, |vertexStage|)
+        1. For every |vertexBuffer| in |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}},
+            the sum of every |vertexBuffer|.{{GPUVertexBufferLayoutDescriptor/attributes}}.length combined
+            is less than or equal to {{GPULimits/maxVertexAttributes|GPULimits.maxVertexAttributes}}.
```
Comment (kainino0x) [inline spec/index.bs:4222]:
```suggestion
        1. The sum of |vertexBuffer|.{{GPUVertexBufferLayoutDescriptor/attributes}}.length,
            over every |vertexBuffer| in |descriptor|.{{GPUVertexStateDescriptor/vertexBuffers}},
            is less than or equal to {{GPULimits/maxVertexAttributes|GPULimits.maxVertexAttributes}}.
```

**Review by kainino0x — APPROVED** [review]

**Pull request merged**